src/lib/transition_router/transition_router.ml

```diff
@@ -66,32 +66,88 @@ module Make (Inputs : Inputs_intf) :
+        ( bootstrap_transition
+        , Strict_pipe.drop_head Strict_pipe.buffered
+        , unit )
+        Strict_pipe.Writer.t }
+
+  let kill reader writer =
+    Strict_pipe.Reader.clear reader ;
+    Strict_pipe.Writer.close writer
+
+  (* TODO: Wrap frontier with a Mvar #1323 *)
   let run ~logger ~network ~time_controller ~frontier ~ledger_db
       ~transition_reader =
-    let clear_reader, clear_writer = Strict_pipe.create Synchronous in
-    let tfc_reader, tfc_writer =
-      Strict_pipe.create (Buffered (`Capacity 10, `Overflow Drop_head))
-    in
-    let bootstrap_controller_reader, bootstrap_controller_writer =
-      Strict_pipe.create (Buffered (`Capacity 10, `Overflow Drop_head))
-    in
-    let is_bootstrapping = ref false in
-    let ancestor_prover =
-      Ancestor.Prover.create ~max_size:(2 * Transition_frontier.max_length)
-    in
     let start_bootstrap (`Transition incoming_transition, `Time_received tm) =
-      Strict_pipe.Writer.write clear_writer `Clear |> don't_wait_for ;
-      is_bootstrapping := true ;
+      let ancestor_prover =
+        Ancestor.Prover.create ~max_size:(2 * Transition_frontier.max_length)
+      in
+      let bootstrap_controller_reader, bootstrap_controller_writer =
+        create_bufferred_pipe ()
+      in
       Strict_pipe.Writer.write bootstrap_controller_writer
         (`Transition incoming_transition, `Time_received (to_unix_timestamp tm)) ;
       let%map () =
         Bootstrap_controller.run ~parent_log:logger ~network ~ledger_db
           ~ancestor_prover ~frontier
           ~transition_reader:bootstrap_controller_reader
       in
-      is_bootstrapping := false
+      (bootstrap_controller_reader, bootstrap_controller_writer)
+    in
+    let start_tfc ~verified_transition_writer ~clear_reader =
+      let transition_reader, transition_writer = create_bufferred_pipe () in
+      let new_verified_transition_reader =
+        Transition_frontier_controller.run ~logger ~network ~time_controller
+          ~frontier ~transition_reader ~clear_reader
+      in
+      Strict_pipe.Reader.iter new_verified_transition_reader
+        ~f:
+          (Fn.compose Deferred.return
+             (Strict_pipe.Writer.write verified_transition_writer))
+      |> don't_wait_for ;
+      (transition_reader, transition_writer)
     in
-    Strict_pipe.Reader.iter transition_reader ~f:(fun network_transition ->
+    let clear_reader, clear_writer = Strict_pipe.create Synchronous in
+    let verified_transition_reader, verified_transition_writer =
+      create_bufferred_pipe ()
+    in
+    let tfc_reader, tfc_writer =
+      start_tfc ~verified_transition_writer ~clear_reader
+    in
+    let bootstrap_controller_reader, bootstrap_controller_writer =
+      create_bufferred_pipe ()
+    in
+    let init =
+      { tfc_reader
+      ; tfc_writer
+      ; bootstrap_controller_reader
+      ; bootstrap_controller_writer }
+    in
+    Strict_pipe.Reader.fold transition_reader ~init
```

Comment:
does this need to be a fold? it doesn't look like you're changing the `pipe_acc`

Reply:
There is no fold anymore. I made a clear distinction of what mode you are in